File: src/main/scala/problems/Problem_0003.scala
package problems

import math.{ floor, sqrt }

/**
 * The prime factors of 13195 are 5, 7, 13 and 29. What is the largest prime
 * factor of the number 600851475143 ?
 */
object Problem_0003 extends Problem {

  def isPrime(n: Long): Boolean = n match {
    case 1 => false
    case 2 => true
    case _ => {
      lazy val maxFactor = floor(sqrt(n)).toInt
      2 #:: (3 to maxFactor by 2 toStream) forall (n % _ != 0)
    }
  }

  def answer = {
    val n = 600851475143l
    lazy val maxFactor = floor(sqrt(n)).toInt

    (2 :: (3 to maxFactor by 2 toList) filter (n % _ == 0) filter (isPrime(_))).max
  }

}

File: src/main/java/org/projecteuler/Problem_0003.java
package org.projecteuler;

import java.util.Stack;

/**
 * The prime factors of 13195 are 5, 7, 13 and 29. What is the largest prime
 * factor of the number 600851475143 ?
 */
public class Problem_0003 {

    private static final long num = 600851475143L;

    /**
     * Its 01:21 on a Saturday, this is going to be "good"
     * Could not figure out why the initial version was talking so long,
     * was calculating for all numbers where div < num / 2 instead of sqrt(num) O_o
     *
     */
    private static long solve() {
        Stack<Long> factors = determineFactors(num);

        while (!factors.empty()) {
            long factor = factors.pop();

            if (isPrime(factor)) {
                return factor;
            }
        }

        return 1;
    }

    private static Stack<Long> determineFactors(long num) {
        Stack<Long> factors = new Stack<Long>();

        if (num % 2 == 0) {
            factors.add(2L);
        }

        long devisor = 3;

        while (devisor < (Math.sqrt(num))) {
            devisor += 2;

            if (num % devisor == 0) {
                factors.push(devisor);
            }
        }

        return factors;
    }

    private static boolean isPrime(long num) {
        long divisor = 1;

        if (num % 2 == 0) {
            return false;
        } else {
            while (divisor < (Math.sqrt(num))) {
                divisor += 2;

                if (num % divisor == 0) {
                    return false;
                }
            }

            return true;
        }
    }

    public static void main(String... args) {
        System.out.println(solve());
    }
}
